device_detection.go

```diff
@@ -0,0 +1,192 @@
+package main
+
+import (
+	"net/http"
+	"regexp"
+	"strings"
+)
+
+// DeviceCapabilities represents the streaming capabilities of a device
+type DeviceCapabilities struct {
+	SupportsHLS    bool
+	IsMobile       bool
+	IsIOS          bool
+	IsAndroid      bool
+	UserAgent      string
+	PreferredCodec string
+}
+
+// iOS user agent patterns for detection - prioritizing User Agent string
+var iosPatterns = []*regexp.Regexp{
+	// Primary iOS device patterns (prioritized)
+	regexp.MustCompile(`(?i)iphone`),
+	regexp.MustCompile(`(?i)ipad`),
+	regexp.MustCompile(`(?i)ipod`),
+	// macOS Safari patterns - also supports HLS natively
+	regexp.MustCompile(`(?i)macintosh.*safari.*version`),
+	regexp.MustCompile(`(?i)mac os x.*safari.*version`),
+}
+
+// Android user agent patterns
+var androidPatterns = []*regexp.Regexp{
+	regexp.MustCompile(`(?i)android`),
+}
+
+// Mobile device patterns
+var mobilePatterns = []*regexp.Regexp{
+	regexp.MustCompile(`(?i)mobile`),
+	regexp.MustCompile(`(?i)tablet`),
+	regexp.MustCompile(`(?i)iphone`),
+	regexp.MustCompile(`(?i)ipad`),
+	regexp.MustCompile(`(?i)ipod`),
+	regexp.MustCompile(`(?i)android`),
+	regexp.MustCompile(`(?i)blackberry`),
+	regexp.MustCompile(`(?i)windows phone`),
+}
+
+// DetectDeviceCapabilities analyzes the HTTP request to determine device capabilities
+// Prioritizes User Agent string detection as requested
+func DetectDeviceCapabilities(r *http.Request) DeviceCapabilities {
+	if r == nil {
+		return DeviceCapabilities{
+			SupportsHLS:    false,
+			IsMobile:       false,
+			IsIOS:          false,
```

Comment:
Is this possible, and relevant, to have some kind of `Enum` here ? Like `enum OS { IOS, Android, Linux, Windows, BSD, ..., Unknown; }`

Reply:
Not really? The decisions we need to make from these properties are: 

* hls support level (iOS natively and Android via hls.js) vs flv (desktop browsers via mpegts.js)
* Bandwidth optimization (reduced resolution and chunk buffering on mobile devices) 

Linux, Windows, etc don't really factor into the equation.